pkg/ruler/scheduler.go

```diff
@@ -182,6 +184,18 @@ func (s *scheduler) poll() (map[string]configs.VersionedRulesConfig, error) {
 func (s *scheduler) addNewConfigs(now time.Time, cfgs map[string]configs.VersionedRulesConfig) {
 	// TODO: instrument how many configs we have, both valid & invalid.
 	level.Debug(util.Logger).Log("msg", "adding configurations", "num_configs", len(cfgs))
+	cycleNanos := float64(s.evaluationInterval.Nanoseconds())
+	nextEvalCycle := time.Unix(0, int64(math.Ceil(float64(now.UnixNano())/cycleNanos)*cycleNanos))
```

Comment:
I don't understand this.

 1. I don't get the purpose, what it's trying to achieve
 2. I can't tell how it's meaningfully different from `time.Unix(0, now.UnixNano())`

Reply:
It's trying to compute the start of the next evaluation cycle, so that we can distribute the load independently of when the ruler started, and evaluations per instance are consistently spaced across ruler restarts.

If now is 20 and the evaluation cycle is 15; we want to compute 30 as the start of the next evaluation cycle; `20 / 15 * 15 == 20`, but `ceil(20 / 15) * 15 == 30`

Reply:
It occurs to me that if now is 1, then this will delay all rule evaluation until 15, rather than starting those evaluations which land sooner (e.g. at 2-14). I'll tweak it to avoid waiting once we settle the other questions.

Reply:
Thanks for the explanation. Looking forward to your tweak.